### PR TITLE
flask_iiif: iiif signals and decorators

### PR DESCRIPTION
**Pull request by drjova**
* NEW adds signals in the restful api.

* NEW adds a configurable decorator to the api.

Signed-off-by: Harris Tzovanakis <me@drjova.com>

### CODE REVIEW

File: flask_iiif/restful.py [hunk 1]
```diff
@@ -68,22 +56,31 @@ class IIIFImageAPI(Resource):
...
             version=version,
+            uuid=uuid,
             region=region,
             size=size,
             rotate=rotation,
             quality=quality,
             image_format=image_format
         )
+        # Trigger event before proccess the api request
+        iiif_before_process_request.send(api_parameters)
```
Comment (jirikuncar) [inline flask_iiif/restful.py]: the first argument (`sender`) should be for sure `self` in order to have access to correct object in signal handler.

File: flask_iiif/decorators.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+            abort(500, message=error.message, code=500)
+        except IIIFValidatorError as error:
+            abort(400, message=error.message, code=400)
+        except (MultimediaError, MultimediaImageNotFound,
+                MultimediaImageForbidden) as error:
+            abort(error.code, message=error.message, code=error.code)
+    return inner
+
+
+def api_decorator(f):
```
Comment (jirikuncar) [inline flask_iiif/decorators.py]: Can you explain what does it do?

File: flask_iiif/decorators.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+from .errors import (
+    IIIFValidatorError, MultimediaError, MultimediaImageForbidden,
+    MultimediaImageFormatError, MultimediaImageNotFound,
+    MultimediaImageQualityError, MultimediaImageRotateError,
+    MultmediaImageCropError, MultmediaImageResizeError
+)
+
+__all__ = ('api_decorator', 'error_handler', )
+
+_iiif = LocalProxy(lambda: current_app.extensions['iiif'])
```
Comment (jirikuncar) [inline flask_iiif/decorators.py]: in order to have consistent naming I would go for `current_iiif` if you really need this proxy.

File: flask_iiif/signals.py [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+# more details.
+
+"""Flask IIIF Signals."""
+
+from blinker import Namespace
+
+_signals = Namespace()
+
+# Before request
+iiif_before_process_request = _signals.signal('iiif-before-process-request')
```
Comment (jirikuncar) [inline flask_iiif/signals.py]: Please document what data are sent in the signal.

Example: https://github.com/inveniosoftware/invenio/blob/master/invenio/modules/accounts/signals.py

Comment (tiborsimko) [inline flask_iiif/signals.py]: Also, in the commit log message, "NEW adds signals in the restful api" would produce release note message that wouldn't be very informative to prospective users.  Perhaps expand the phrase little a bit, like: "Adds new signals to REST API that permit to do foo and bar."

Comment (drjova) [inline flask_iiif/signals.py]: Done, let me know if it's detailed enough thanks!

```
    flask_iiif: iiif signals and decorators

    * NEW adds new signals to REST API that permits to have access before
      and after process of the request as well as after the validation of
      IIIF.

    * NEW adds a configurable decorator to the REST API which can be
      configure with the `api_decorator_handler`.

    * NOTE with the new decorator can be used to restrict access to the REST
      API.
```

Comment (jirikuncar) [inline flask_iiif/signals.py]: @drjova it would be nice to start a sentence after label with capital letter.

**Comment by jirikuncar** [comment]
@drjova can you rephrase the last bullet `* NOTE With the new decorator can be used to restrict access to the REST API.`?

**Comment by drjova** [comment]
@jirikuncar done

```
    flask_iiif: iiif signals and decorators

    * NEW Adds new signals to REST API that permits to have access before
      and after process of the request as well as after the validation of
      IIIF.

    * NEW Adds a configurable decorator to the REST API which can be
      configure with the `api_decorator_handler`.

    * NOTE The decorator can be used to restrict access to the REST
      API.

    Signed-off-by: Harris Tzovanakis <me@drjova.com>

```

File: tests/test_signals_and_deocorators.py [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+
+            @iiif_after_validation.connect_via(self.app)
+            def on_after_validation(uuid):
+                self.assertEqual(
+                    uuid,
+                    'j.a.r.v.i.s.'
+                )
+
+            # fire the signals
+            iiif_before_process_request.send('ironman-mark-48')
```
Comment (jirikuncar) [inline tests/test_signals_and_deocorators.py]: You should actually execute the REST API request.

File: tests/test_signals_and_deocorators.py [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+                )
+
+            @iiif_after_process_request.connect_via(self.app)
+            def on_after_process(uuid):
+                self.assertEqual(
+                    uuid,
+                    'ironman-mark-43'
+                )
+
+            @iiif_after_validation.connect_via(self.app)
```
Comment (jirikuncar) [inline tests/test_signals_and_deocorators.py]: The `connect_via` argument should be `iiif` (`current_iiif.get_current_object()`).

File: tests/test_signals_and_deocorators.py [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+
+            def hydra():
+                abort(500)
+
+            _iiif.api_decorator_handler(hydra)
+
+            self.assertRaises(
+                InternalServerError,
+                hydra
+            )
```
Comment (jirikuncar) [inline tests/test_signals_and_deocorators.py]: you should test the integration with the REST API here.

File: tests/test_signals_and_deocorators.py (hunk 5 above)
Comment (drjova) [inline tests/test_signals_and_deocorators.py]: I'll make a separated REST API test to increase the overall coverage

File: tests/test_restful_api_signals.py [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+            size='full',
+            rotation='0',
+            quality='default',
+            image_format='png'
+        )
+
+        self.get(
+            'iiifimageapi',
+            urlargs=expected
+        )
```
Comment (drjova) [inline tests/test_restful_api_signals.py]: @jirikuncar if there is a better way to test `signals` please let me know

**Comment by drjova** [comment]
@jirikuncar is it any way to deactivate `coveralls` comment?

File: tests/test_restful_api_signals.py [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+
+        with signal_listener(iiif_before_process_request) as listener:
+            self.get(
+                'iiifimageapi',
+                urlargs=expected
+            )
+            results = listener.assert_signal()
+            self.assertEqual(
+                results[1].get('uuid'),
```
Comment (jirikuncar) [inline tests/test_restful_api_signals.py]: Will this work?
```python
self.assertEqual(results[1],expected)
```

Comment (drjova) [inline tests/test_restful_api_signals.py]: Now it will work thanks!
